tests/abi/tools/revng-check-compatibility-with-abi/ABIRuntimeTestResultParser.h

```diff
@@ -4,86 +4,137 @@
+struct KeyedObjectTraits<abi::runtime_test::Register> {
   static model::Register::Values
-  key(const abi::artifact::ModelRegister &Object) {
+  key(const abi::runtime_test::Register &Object) {
     return Object.Name;
   }
-  static abi::artifact::ModelRegister
+  static abi::runtime_test::Register
   fromKey(const model::Register::Values &Key) {
     return { Key };
   }
 };
 
-namespace abi::artifact {
+namespace abi::runtime_test {
+
+struct State {
+  SortedVector<Register> Registers;
+  std::vector<std::byte> Stack;
+
+  State(SortedVector<Register> &&Registers, std::vector<std::byte> &&Stack) :
+    Registers(std::move(Registers)), Stack(std::move(Stack)) {}
+
+  std::optional<llvm::ArrayRef<std::byte>>
+  operator[](model::Register::Values Register) const {
+    auto Iterator = Registers.find(Register);
+    if (Iterator != Registers.end())
+      return Iterator->Bytes;
+    else
+      return std::nullopt;
+  }
+};
+
+struct TestState {
+  State StateBeforeTheCall;
+  State StateAfterTheCall;
+  State StateAfterTheReturn;
+};
+
+struct ArgumentTest : public TestState {
+  std::vector<Argument> Arguments;
 
-using ModelRegisters = SortedVector<ModelRegister>;
+  ArgumentTest(TestState &&State, std::vector<Argument> &&Arguments) :
+    TestState(std::move(State)), Arguments(std::move(Arguments)) {}
+};
 
-struct Iteration {
-  ModelRegisters Registers;
-  Stack Stack;
-  Arguments Arguments;
+struct ReturnValueTest : public TestState {
+  RawWord ReturnValueAddress;
   Argument ReturnValue;
+  Argument ExpectedReturnValue;
+
+  ReturnValueTest(TestState &&State,
+                  RawWord &&ReturnValueAddress,
+                  Argument &&ReturnValue,
+                  Argument &&ExpectedReturnValue) :
+    TestState(std::move(State)),
+    ReturnValueAddress(std::move(ReturnValueAddress)),
+    ReturnValue(std::move(ReturnValue)),
+    ExpectedReturnValue(std::move(ExpectedReturnValue)) {}
 };
 
-struct FunctionArtifact {
+template<typename Iteration>
+struct Named {
   llvm::StringRef Name;
-  llvm::SmallVector<Iteration, 5> Iterations = {};
+  std::vector<Iteration> Iterations = {};
+  const std::vector<Iteration> &operator*() const { return Iterations; }
 };
 
-} // namespace abi::artifact
+} // namespace abi::runtime_test
 
-template<>
-struct KeyedObjectTraits<abi::artifact::FunctionArtifact> {
-  static llvm::StringRef key(const abi::artifact::FunctionArtifact &Object) {
-    return Object.Name;
-  }
-  static abi::artifact::FunctionArtifact fromKey(const llvm::StringRef &Key) {
-    return { Key };
-  }
+template<typename Iteration>
+struct KeyedObjectTraits<abi::runtime_test::Named<Iteration>> {
+  using Type = abi::runtime_test::Named<Iteration>;
+  static llvm::StringRef key(const Type &Object) { return Object.Name; }
+  static Type fromKey(const llvm::StringRef &Key) { return { Key }; }
 };
 
-namespace abi::artifact {
+namespace abi::runtime_test {
 
 struct Parsed {
```

Comment:
1. Just call this `TestCase` or something. The fact that its the result of parsing is not relevant to its users.
2. This seems like a pretty plain format what prevents us from making *this* the YAML format?

Reply:
1. Sounds good.
2. The need to redesign it (I _really_ didn't want to add even more). In the early iterations we were _very_ limited as to what we can do when emitting stuff. Now, thanks to the interrupt handlers, the possibilities are limitless.

Reply:
2. OK, leave a TODO.

Reply:
1. I went with `TestedFunctions` since `Case` sounds weird when the data structure contains _all_ the tests.

Reply:
Done.